README
$Id: README 1463 2004-09-08 07:16:39Z comdog $

You can install this using in the usual Perl fashion

	perl Makefile.PL
	make
	make test
	make install
	
The documentation is in the module file.  Once you install
the file, you can read it with perldoc.

	perldoc Object::Iterate
	
If you want to read it before you install it, you can use
perldoc directly on the module file.

	perldoc lib/Iterate.pm
	
This module is also in CVS on SourceForge

	http://sourceforge.net/projects/brian-d-foy/


Enjoy, 

brian d foy, bdfoy@cpan.org

Changes
1.11 - Sun Dec  9 01:08:19 2007
	* Cleanups for the move from CVS to SVN
	* No need to upgrade

1.13 - Mon Oct  7 12:14:25 2013
	* Add Meta data, mess around with distribution file structure
	* no code changes

1.10 - Tue Jan  9 22:38:03 2007
	* updated copyright and license info
	* no code changes, so no need to upgrade

1.09 - Thu Jun  8 06:41:22 2006
	* distro cleanups, bumped version number to stable release.
	No need to upgrade.

0.53 - Sat Jan 28 12:34:59 2006
	* Fixed some typoes, upgraded the Test::Manifest use in Makefile.PL
	* No need to upgrade if you already have Object::Iterate

0.52 - Fri Mar 11 22:24:15 2005
	* Added POD coverage tests, fixed issues
	* No code updates: no need to upgrade

0.51 - Fri Mar 11 22:12:48 2005
	* Added POD coverage tests and fixed those problems
	* No code changes: no need to upgrade

0.51 - Wed Sep  8 02:27:16 2004
	* cleaned up the distro (where'd that MANIFEST hide?)
	* updated the tests and reduced the dependencies
	* no need to upgrade if you already have the module



lib/Iterate.pm
# $Id: Iterate.pm 2457 2007-12-09 07:01:01Z comdog $
package Object::Iterate;
use strict;

use warnings;
no warnings;

use subs qw(_check_object);
use vars qw(@ISA $VERSION @EXPORT_OK %EXPORT_TAGS
	$Next $More $Init $Final
	);

=head1 NAME

Object::Iterate - iterators for objects that know the next element

=head1 SYNOPSIS

	use Object::Iterate qw(iterate igrep imap);

	iterate {...} $object;

	my @filtered    = igrep {...} $object;

	my @transformed = imap {...} $object;

=head1 DESCRIPTION

This module provides control structures to iterate through the
elements of an object that cannot be represented as list of items all
at once.  Objects can represent a virtual collection that is beyond
the reaches of foreach, map, and grep because they cannot turn
themselves into a list.

If the object can return a next element, it can use this module.
Iterate assumes that the object responds to C<__next__> with the next
element, and to C<__more__> with TRUE or FALSE if more elements remain
to be processed.  The C<__init__> method is called before the first
iteration (if it exists), and is silently skipped otherwise. The
control structure continues until the C<__more__> method returns FALSE
(which does not mean that it visited all of the elements but that the
object has decided to stop iterating).  At the end of all iterations
(when C<__more__> returns false), C<Object::Iterate> calls
C<__final__> if it exists, and skips it otherwise.

Each control structure sets C<$_> to the current element, just like
foreach, map, and grep.

=head2  Mutable method names

You do not really have to use the C<__next__>, C<__more__>,
C<__init__>, or C<__final__> names. They are just the defaults which
<Object::Iterate> stores in the package variables C<$Next>, C<$More>,
C<$Init>, and C<$Final> respectively.  This module does not export
these variables, so you need to use the full package specification to
change them (I<i.e.> C<$Object::Iterate::$Next>). If your object does
not have the specified methods, the functions will die.  You may want
to wrap them in eval blocks.

Since this module uses package variables to storethese methods names,
the method names apply to every use of the functions no matter the
object.  You might want to local()-ise the variables for different
objects.

Before any control structure does its job, it checks the object to see
if it can respond to these two methods, whatever you decide to call
them, so your object must know that it can respond to these methods. 
AUTOLOADed methods cannot work since the module cannot know if they
exist.

=cut

use Carp qw(croak);
use Exporter;

@ISA         = qw(Exporter);
@EXPORT_OK   = qw(iterate igrep imap);
$VERSION     = 1.11;

%EXPORT_TAGS = (
	all => \@EXPORT_OK,
	);

$Next  = '__next__';
$More  = '__more__';
$Init  = '__init__';
$Final = '__final__';

sub _check_object
	{
	croak( "iterate object has no $Next() method" )
		unless UNIVERSAL::can( $_[0], $Next );
	croak( "iterate object has no $More() method" )
		unless UNIVERSAL::can( $_[0], $More );

	$_[0]->$Init if UNIVERSAL::can( $_[0], $Init );

	return 1;
	}

=over 4

=item iterate BLOCK, OBJECT

Applies BLOCK to each item returned by C<OBJECT->__next__>.

	iterate { print "$_\n" } $object;

This is the same thing as using a while loop, but C<iterate()>
stays out of your way.

	while( $object->__more__ )
		{
		local $_ = $object->__next__;

		...BLOCK...
		}
=cut

sub iterate (&$)
	{
	my $sub    = shift;
	my $object = shift;

	_check_object( $object );

	while( $object->$More )
		{
		local $_;

		$_ = $object->$Next;

		$sub->();
		}

	$object->$Final if $object->can( $Final );
	}

=item igrep BLOCK, OBJECT

Applies BLOCK to each item returned by C<OBJECT->__next__>, and returns
all of the elements for which the BLOCK returns TRUE.

	my $output = igrep { print "$_\n" } $object;

This is a grep for something that cannot be represented as a
list at one time.

	while( $object->__more__ )
		{
		local $_ = $object->__next__;

		push @output, $_ if ...BLOCK...;
		}

=cut

sub igrep (&$)
	{
	my $sub    = shift;
	my $object = shift;

	_check_object( $object );

	my @output = ();

	while( $object->$More )
		{
		local $_;

		$_ = $object->$Next;

		push @output, $_ if $sub->();
		}

	$object->$Final if $object->can( $Final );

	wantarray ? @output : scalar @output;
	}

=item imap BLOCK, OBJECT

Applies BLOCK to each item returned by C<OBJECT->__next__>, and returns
the combined lists that BLOCK returns for each of the elements.

	my $output = imap { print "$_\n" } $object;

This is a map for something that cannot be represented as a
list at one time.

	while( $object->$More )
		{
		local $_ = $object->__next__;

		push @output, ...BLOCK...;
		}

=cut

sub imap (&$)
	{
	my $sub    = shift;
	my $object = shift;

	_check_object( $object );

	my @output = ();

	while( $object->$More )
		{
		local $_;

		$_ = $object->$Next;

		push @output, $sub->();
		}

	$object->$Final if $object->can( $Final );

	@output;
	}

=back

=head1 ERROR MESSAGES

=over 4

=item iterate object has no C<__more__()> method at script line N

You need to provide the method to let C<Object::Iterate> determine if
more elements are available.  You don't have to call it C<__more__> if you
change the value of C<$Object::Iterate::More>.

=item iterate object has no C<__next__()> method at script line N

You need to provide the method to let Object::Iterate fetch the next
element.  You don't have to call it C<__next__> if you change the value of
C<$Object::Iterate::Next>.

=back

=head1 SOURCE AVAILABILITY

This source is part of a SourceForge project which always has the
latest sources in CVS, as well as all of the previous releases.

	http://sourceforge.net/projects/brian-d-foy/

If, for some reason, I disappear from the world, one of the other
members of the project can shepherd this module appropriately.

=head1 TO DO

* let the methods discover the method names per object.

=head1 CREDITS

Thanks to Slaven Rezic for adding C<__init__> support

=head1 AUTHOR

brian d foy, C<< <bdfoy@cpan.org> >>.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2002-2007 brian d foy.  All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


Makefile.PL
# $Id: Makefile.PL 2457 2007-12-09 07:01:01Z comdog $
use ExtUtils::MakeMaker;

require 5.006;

eval "use Test::Manifest 1.14";

WriteMakefile
	(
	'NAME'          => 'Object::Iterate',
	'ABSTRACT'      => 'iterators for objects that know the next element',
	'AUTHOR'        => 'brian d foy <bdfoy@cpan.org>',
	'VERSION_FROM'  => 'lib/Iterate.pm',
	'LICENSE'       => 'perl',

	'PM'            => {
		'lib/Iterate.pm'    => '$(INST_LIBDIR)/Iterate.pm',
		'lib/Tester.pm'     => '$(INST_LIBDIR)/Iterate/Tester.pm',
		},

	'PREREQ_PM' => {
		'Test::Data'       => '0',
		'Test::More'       => '0',
		},

	'MAN3PODS'     => {
		'lib/Iterate.pm' => '$(INST_MAN3DIR)/Object::Iterate.3',
		'lib/Tester.pm'  => '$(INST_MAN3DIR)/Object::Iterate::Tester.3',
		},

	clean => { FILES => 'Object-Iterate-*' },
	);
